### PR TITLE
Add transaction model aiopg

### PR DESCRIPTION
**Pull request by vir-mir**
Todo task
- [x] three isolation level [pg](https://postgrespro.com/docs/postgrespro/9.5/transaction-iso.html)
- [x] Nested transactions [pg](https://postgrespro.com/docs/postgrespro/9.5/sql-savepoint.html)
- [x] exemples new style
- [x] exemples old style
- [x] finalization warning
- [x] test cover
- [ ] docs
- [x] add ``__slots__``
- [x] fix `aiopg/transaction.py:21`

### CODE REVIEW

**Comment by jettify** [comment]
Just for your information, you can put python3.5 tests into `tests/pep492` folder to avoid syntax errors on older python versions.

**Comment by asvetlov** [comment]
@vir-mir would you add missing tests?

**Comment by vir-mir** [comment]
@asvetlov hi! Yes.
I will finish, remove from the WIP (Work in process) title.

**Comment by vir-mir** [comment]
@asvetlov @jettify Done

sorry, I  did not write the documentation.
But I added `exemples/transactio.py` and `exemples/transactio_oldstyle.py`

**Comment by vir-mir** [comment]
LGTM?

**Comment by asvetlov** [comment]
please wait a bit

**Comment by asvetlov** [comment]
@vir-mir we should decide: support typing annotations for the whole project or not.
I like it (while infrastructure is not really ready yet from my perspective).

If we will decide to support -- we should add annotations to every public `aiopg` method and check it by `mypy` tool strictly.
If not -- please don't add not checked annotations at all.
Would you be a volunteer for the task?

**Comment by jettify** [comment]
Exactly my thoughts, should we support external file like `yarl` or something inplace like here?

**Comment by jettify** [comment]
I believe we will add annotation eventually anyway, but lets drop it from this PR and add them separately  with `mypy` supervision

File: aiopg/transaction.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+    def __call__(self, readonly, deferrable):
+        return self.value(readonly, deferrable)
+
+
+class Transaction:
+    __slots__ = ('_cur', '_is_begin', '_isolation', '_unique_id')
+
+    def __init__(self, cur, isolation_level: IsolationLevel,
+                 readonly: bool = False, deferrable: bool = False):
```
Comment (vir-mir) [inline aiopg/transaction.py]: Good

**Comment by vir-mir** [comment]
Hi.
I propose to put the default isolation `Read Committed` 
I add `begin` and `begin_nested` probably tomorrow.

**Comment by jettify** [comment]
`Read Committed` is default for PG so it is reasonable choice.

**Comment by vir-mir** [comment]
@asvetlov  I'm add factories in cursor and drop spaces around

example

```python
async def begin(engine):
    async with engine.acquire() as cur:
        async with cur.begin():
            await cur.execute("insert into tbl values(1, 'data')")

        async with cur.begin():
            resp = await cur.execute('select * from tbl')
            row = await resp.fetchall()
            assert row == [(22, 'read only'), (1, 'data'), ]


async def begin_nested(engine):
    async with engine.acquire() as cur:
        async with cur.begin_nested():
            await cur.execute("insert into tbl values(1, 'data')")

            try:
                async with cur.begin_nested():
                    await cur.execute("insert into tbl values(1/0, 'no data')")
            except psycopg2.DataError:
                    pass

            async with cur.begin_nested():
                await cur.execute("insert into tbl values(2, 'data')")

        async with cur.begin_nested():
            resp = await cur.execute('select * from tbl')
            row = await resp.fetchall()
            assert row == [(22, 'read only'), (1, 'data'),  (2, 'data'), ]
```

**Comment by vir-mir** [comment]
resolve conflicts

@asvetlov any news?

**Comment by drednout** [comment]
Hello, everyone!
What is the actual status of this PR? Is there any chances to have transaction functionality for raw SQL in aiopg upstream?

**Comment by vir-mir** [comment]
Dear all!
I rebuilt, I had to do push a force. Sorry. I also pull the master and corrected all the comments.

@asvetlov @jettify  review, please

File: aiopg/cursor.py [hunk 2]
```diff
@@ -146,6 +147,16 @@ def callproc(self, procname, parameters=None, *, timeout=None):
         else:
             yield from self._conn._poll(waiter, timeout)
 
+    def begin(self):
+        return _TransactionBeginContextManager(self._transaction.begin())
```
Comment (asvetlov) [inline aiopg/cursor.py]: Not covered

File: aiopg/cursor.py [hunk 3]
```diff
@@ -146,6 +147,16 @@ def callproc(self, procname, parameters=None, *, timeout=None):
...
+
+    def begin_nested(self):
+        if not self._transaction.is_begin:
```
Comment (asvetlov) [inline aiopg/cursor.py]: Not called by tests

File: aiopg/utils.py [hunk 4]
```diff
@@ -106,8 +102,33 @@ def __aexit__(self, exc_type, exc, tb):
...
+
+            self._obj = None
 
+
+class _TransactionBeginContextManager(_ContextManager):
+    if PY_35:
+
+        @asyncio.coroutine
+        def __aexit__(self, exc_type, exc_val, exc_tb):
+            if exc_type is not None:
```
Comment (asvetlov) [inline aiopg/utils.py]: Not covered by tests

File: aiopg/cursor.py (hunk 2 above)
Comment (vir-mir) [reply aiopg/cursor.py]: done

File: aiopg/cursor.py (hunk 3 above)
Comment (vir-mir) [reply aiopg/cursor.py]: done

File: aiopg/utils.py (hunk 4 above)
Comment (vir-mir) [reply aiopg/utils.py]: done

**Comment by asvetlov** [comment]
Cool!
The code looks perfect but lacks documentation.

**Comment by jettify** [comment]
Lets do docs in separate PR, this one already large.

**Comment by asvetlov** [comment]
> Lets do docs in separate PR, this one already large.
Ok. @vir-mir please merge.

Documentation is a show stopper for next release.

**Comment by vir-mir** [comment]
@jettify @asvetlov  I added code samples for the cursor. 
docs of the class of transactions, I propose to pass in another PR

https://github.com/aio-libs/aiopg/pull/219/commits/96cf79834a1a41c968b0b354253ead5221d6e9b9

**Pull request merged**